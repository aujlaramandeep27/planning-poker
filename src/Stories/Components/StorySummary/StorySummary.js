import React, {useEffect} from 'react';
import {connect} from 'react-redux';
import {Page} from '../../../Common/Page';
import {viewStories} from '../../../Navigation/route-actions';
import {UpdateCurrentStory} from "../../../Meetings/Actions/CurrentMeetingActions";
import {StartPollingStoryEstimates, StopPollingStoryEstimates} from "../../Actions/StoryEstimatesActions";

export function StorySummary({currentMeeting, currentStory, storyEstimates, goToStories, updateCurrentStory, startPollingStoryEstimates, stopPollingStoryEstimates, match}) {

    useEffect(
        () => {
            updateCurrentStory(match.params.meetingId, match.params.storyId);
            startPollingStoryEstimates(match.params.storyId);
            return () => {
                updateCurrentStory(match.params.meetingId, null);
                stopPollingStoryEstimates();
            };
        },
        [updateCurrentStory, startPollingStoryEstimates, stopPollingStoryEstimates, match],
    );

    const sum = (a, b) => a + b;

    function average(numbers) {
        return numbers && sum ? Math.ceil(numbers.reduce(sum, 0) / numbers.length) : 0;
    }

    function hasValidEstimate(estimate) {
        return estimate.user && estimate.user.name && estimate.estimate;
    }

    function showAverage() {
        const estimates = storyEstimates.filter(hasValidEstimate);
        return estimates.length > 0 ? (
            <tr className="uk-text-bold" key={'average'}>
                <td className="uk-text-center">Average</td>
                <td className="uk-text-center">{average(estimates.map(estimate => estimate.estimate)).toString()}</td>
            </tr>
        ) : '';
    }

    return (
        <Page title={currentMeeting.name} onBack={() => goToStories(currentMeeting._id)}>
            <dl className="uk-description-list">
                <dt># {currentStory.name} :</dt>
            </dl>
            <div className="uk-align-center uk-width-1-2@m">
                <table className="uk-table uk-table-divider">
                    <thead>
                    <tr>
                        <th className="uk-text-center">Name</th>
                        <th className="uk-text-center">Estimate</th>
                    </tr>
                    </thead>
                    <tbody>
                    {storyEstimates.filter(hasValidEstimate).map((estimate, index) => {
                        return (
                            <tr key={index}>
                                <td className="uk-text-center">{estimate.user.name}</td>
                                <td className="uk-text-center">{estimate.estimate}</td>
                            </tr>
                        );
                    })}
                    {showAverage()}
                    </tbody>
                </table>
            </div>
        </Page>
    );

}

function mapStateToProps(state) {
    return {
        currentMeeting: state.currentMeeting,
        currentStory: state.currentStory,
        storyEstimates: state.storyEstimates
    }
}

function mapDispatchToProps(dispatch) {
    return {
        goToStories: (meetingId) => dispatch(viewStories(meetingId)),
        updateCurrentStory: (meetingId, storyId) => dispatch(UpdateCurrentStory(meetingId, storyId)),
        startPollingStoryEstimates: (storyId) => dispatch(StartPollingStoryEstimates(storyId)),
        stopPollingStoryEstimates: () => dispatch(StopPollingStoryEstimates())
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(StorySummary)