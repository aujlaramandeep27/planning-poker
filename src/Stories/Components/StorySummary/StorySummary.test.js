import {StorySummary} from "./StorySummary";
import {cleanup, render} from "@testing-library/react";
import React from "react";

describe('Container: StorySummary', () => {

    afterEach(cleanup);

    it('should render correctly with estimates and names', () => {
        const props = {
            currentMeeting: jest.fn().mockReturnValue({
                name: 'Mock Meeting Name'
            }),
            storyEstimates: [
                {
                    user: {
                        name: 'Bob'
                    },
                    estimate: 1
                },
                {
                    user: {
                        name: 'Andy'
                    },
                    estimate: 5
                },
                {
                    user: {
                        name: 'Joe'
                    },
                    estimate: 3
                },
                {
                    user: {
                        name: null
                    },
                    estimate: 'bad data'
                }
            ],
            currentStory: {
                name: 'Mock Story Name'
            },
            goToStories: jest.fn(),
            updateCurrentStory: jest.fn(),
            startPollingStoryEstimates: jest.fn(),
            stopPollingStoryEstimates: jest.fn()
        };
        const {container} = render(<StorySummary  {...props} match={{'params': {'storyId': 'Mock story id'}}}/>);
        expect(container).toMatchSnapshot();
    });
});