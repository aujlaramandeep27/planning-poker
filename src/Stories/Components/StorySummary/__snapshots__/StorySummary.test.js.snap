// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container: StorySummary should render correctly with estimates and names 1`] = `
<div>
  <div
    class="default-background"
  >
    <div
      class="uk-container uk-text-center@m"
    >
      <div
        class="uk-section uk-section-xsmall"
      >
        <span
          class="uk-button uk-button-secondary uk-button-small uk-position-small uk-position-top-left pp-back-arrow"
          data-uk-icon="icon: arrow-left; ratio: 1.25"
        />
      </div>
      <h1
        class="uk-heading-divider uk-margin-top"
      >
        mockConstructor
      </h1>
      <div
        class="uk-align-center uk-width-1-2@m"
      >
        <dl
          class="uk-description-list"
        >
          <dt>
            # 
            Mock Story Name
             :
          </dt>
        </dl>
        <div
          class="uk-align-center uk-width-1-2@m"
        >
          <table
            class="uk-table uk-table-divider"
          >
            <thead>
              <tr>
                <th
                  class="uk-text-center"
                >
                  Name
                </th>
                <th
                  class="uk-text-center"
                >
                  Estimate
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td
                  class="uk-text-center"
                >
                  Bob
                </td>
                <td
                  class="uk-text-center"
                >
                  1
                </td>
              </tr>
              <tr>
                <td
                  class="uk-text-center"
                >
                  Andy
                </td>
                <td
                  class="uk-text-center"
                >
                  5
                </td>
              </tr>
              <tr>
                <td
                  class="uk-text-center"
                >
                  Joe
                </td>
                <td
                  class="uk-text-center"
                >
                  3
                </td>
              </tr>
              <tr
                class="uk-text-bold"
              >
                <td
                  class="uk-text-center"
                >
                  Average
                </td>
                <td
                  class="uk-text-center"
                >
                  3
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
